### PR TITLE
Initial implementation of tuples in type checking

### PR DESCRIPTION
**Pull request by joshuawarner32**
This leaves in place a bunch of TODOs and likely many bugs - notably, I haven't tested codegen/layout at all here.

### CODE REVIEW

File: crates/compiler/constrain/src/expr.rs [hunk 1]
```diff
@@ -252,7 +252,52 @@ pub fn constrain_expr(
                 constraints.exists(field_vars, and_constraint)
             }
         }
-        Update {
+        Expr::Tuple { tuple_var, elems } => {
+            let mut elem_types = Vec::with_capacity(elems.len());
+            let mut elem_vars = Vec::with_capacity(elems.len());
+
+            // Constraints need capacity for each elem
+            // + 1 for the record itself + 1 for record var
```
Comment (ayazhafiz) [inline crates/compiler/constrain/src/expr.rs:260]: s/record/tuple

File: crates/compiler/constrain/src/expr.rs [hunk 2]
```diff
@@ -252,7 +252,52 @@ pub fn constrain_expr(
...
+            let mut rec_constraints = Vec::with_capacity(2 + elems.len());
```
Comment (ayazhafiz) [inline crates/compiler/constrain/src/expr.rs:261]: should this be named tuple_constraints?

File: crates/compiler/constrain/src/expr.rs [hunk 3]
```diff
@@ -252,7 +252,52 @@ pub fn constrain_expr(
...
+
+            let tuple_type = {
+                let typ = types.from_old_type(&Type::Tuple(elem_types, TypeExtension::Closed));
+                constraints.push_type(types, typ)
+            };
+
+            let tuple_con = constraints.equal_types_with_storage(
+                tuple_type,
+                expected,
+                Category::Record,
```
Comment (ayazhafiz) [inline crates/compiler/constrain/src/expr.rs:289]: let's add a Category::Tuple for error reporting purposes

File: crates/compiler/unify/src/unify.rs [hunk 4]
```diff
@@ -2008,6 +2013,118 @@ fn unify_record<M: MetaCollector>(
...
+) -> Outcome<M> {
+    let subs = &mut env.subs;
+
+    let (separate, ext1, ext2) = separate_tuple_elems(subs, elems1, ext1, elems2, ext2);
+
+    let shared_elems = separate.in_both;
+
+    if separate.only_in_1.is_empty() {
+        if separate.only_in_2.is_empty() {
+            // these variable will be the empty record, but we must still unify them
```
Comment (ayazhafiz) [inline crates/compiler/unify/src/unify.rs:2034]: s/record/tuple

File: crates/compiler/unify/src/unify.rs [hunk 5]
```diff
@@ -557,6 +557,11 @@ fn debug_print_unified_types<M: MetaCollector>(
             SubsFmtContent(&content_2, env.subs),
         );
 
+        if ctx.second.index() == 1357 {
+            eprintln!("BLARG {:?}", content_1);
+            eprintln!("BLARG {:?}", content_2);
+        }
```
Comment (ayazhafiz) [inline crates/compiler/unify/src/unify.rs:563]: should this be removed?

File: crates/compiler/types/src/types.rs [hunk 6]
```diff
@@ -3420,6 +3622,13 @@ pub struct RecordStructure {
     pub ext: Variable,
 }
 
+#[derive(Debug)]
+pub struct TupleStructure {
+    /// Invariant: these should be sorted!
+    pub elems: Vec<(usize, Variable)>,
```
Comment (ayazhafiz) [inline crates/compiler/types/src/types.rs:3628]: there is a data structure VecMap in roc_collections that might be better suited here to help ensure order and no duplicates

File: crates/compiler/types/src/subs.rs [hunk 7]
```diff
@@ -4707,6 +4987,34 @@ fn storage_copy_var_to_help(env: &mut StorageCopyVarToEnv<'_>, var: Variable) ->
...
+                        let it = (new_variables.indices()).zip(elems.iter_variables());
+                        for (target_index, var_index) in it {
+                            let var = env.source[var_index];
+                            let copy_var = storage_copy_var_to_help(env, var);
+                            env.target.variables[target_index] = copy_var;
+                        }
+
+                        let elem_index_start = env.target.tuple_elem_indices.len() as u32;
+
+                        // TODO: introduce a DenseTupleElems type, so we don't have to
```
Comment (ayazhafiz) [inline crates/compiler/types/src/subs.rs:5004]: not needed now but you could encode this pretty easily on the current representation. make tuple_elem_indices_start a result; in the left case, it is the start of the tuple elems indices, in the right case it's the number of indices

File: crates/compiler/solve/tests/solve_expr.rs [hunk 8]
```diff
@@ -1469,6 +1469,21 @@ mod solve_expr {
...
+    #[test]
+    fn tuple_literal_accessor() {
+        infer_eq("(5, 3.14 ).0", "Num *");
+    }
+
+    #[test]
+    fn tuple_literal_accessor_function() {
+        infer_eq(".0 (5, 3.14 )", "Num *");
+    }
+
```
Comment (ayazhafiz) [inline crates/compiler/solve/tests/solve_expr.rs:1486]: can some tests for the type of just tuples and just tuple accessors be added?

File: crates/compiler/solve/src/ability.rs [hunk 9]
```diff
@@ -702,6 +722,21 @@ trait DerivableVisitor {
...
+                        let descend = Self::visit_tuple(subs, var, elems)?;
+                        if descend.0 {
+                            push_var_slice!(elems.variables());
+                            if !matches!(
+                                subs.get_content_without_compacting(ext),
+                                Content::FlexVar(_) | Content::RigidVar(_)
+                            ) {
+                                // TODO: currently, just we suppose the presence of a flex var may
+                                // include more or less things which we can derive. But, we should
+                                // instead recurse here, and add a `t ~ u | u has Decode` constraint as needed.
```
Comment (ayazhafiz) [inline crates/compiler/solve/src/ability.rs:735]: i don't think this comment is relevant anymore and can be removed

**Review by ayazhafiz — COMMENTED** [review]
General design LGTM. I like the idea of sparse representation by default

File: crates/compiler/types/src/subs.rs (hunk 7 above)
Comment (joshuawarner32) [reply crates/compiler/types/src/subs.rs]: Great idea - I left a comment to that effect and renamed the struct to just `TupleElems`.

File: crates/compiler/solve/tests/solve_expr.rs (hunk 8 above)
Comment (joshuawarner32) [reply crates/compiler/solve/tests/solve_expr.rs:1486]: Added a test for tuple types (that's now pushed).

Then I realized you also asked about tuple accessor - which it turns out don't work off the bat. Need to resolve todo!() :)

Comment (joshuawarner32) [reply crates/compiler/solve/tests/solve_expr.rs:1486]: Ok, done!

File: crates/compiler/solve/tests/solve_expr.rs [hunk 10]
```diff
@@ -1469,6 +1469,34 @@ mod solve_expr {
...
+        infer_eq("(5, 3.14 )", "( Num *, Float * )*");
+    }
+
+    #[test]
+    fn tuple_literal_accessor_ty() {
+        infer_eq(".0", "( a )* -> a");
+        infer_eq(".4", "( _, _, _, _, a )* -> a");
+        infer_eq(".5", "( ... 5 omitted, a )* -> a");
+        infer_eq(".200", "( ... 200 omitted, a )* -> a");
+    }
```
Comment (ayazhafiz) [inline crates/compiler/solve/tests/solve_expr.rs:1498]: Doesn't need to happen now, but in a follow up can we add a test to ensure that tuple accessors are in fact generalized? That is, I'm thinking of a test like

```
get0 = .0

{ a: get0 (1, 2), b: get0 ("a", "b", "c") }
```

to make sure it doesn't error

**Review by ayazhafiz — APPROVED** [review]
This is excellent, congrats @joshuawarner32 !!

**Pull request merged**